Show recent topics as a view.

amanagr opened the pull request:
@timabbott this is ready from the UI / UX point of view, I just need to make this merge-able. Opening for a secondary purpose.

amanagr commented:
@timabbott this is reaady for review.

amanagr commented:
Fixed night theme.

timabbott commented:
One thing I notice is that we should be using simplebar for the scrollbar, so it doesn't look like this on Linux:

![image](https://user-images.githubusercontent.com/2746074/86500524-8da54380-bd46-11ea-9773-b7b1ee7382b6.png)

timabbott commented:
I think we also need some sort of thin side-border, maybe in the same style as we have for the navbars, between the columns, to make this look like a proper widget.

timabbott commented on templates/zerver/app/index.html:
```diff
@@ -92,7 +91,12 @@ <h3>{{ _('Loading...') }}</h3>
+            <div class="recent_topics_container">
+                <div id="recent_topics_table"></div>
+            </div>
+        </div>
```
Structurally, I'd expect this to be inside `column-middle-inner` and `<div class="tab-pane active" id="home">` to have the `home `renamed to `message_feed_container` (your `messages_view`), in this template's hierarchy.

timabbott commented on static/js/recent_topics.js:
```diff
@@ -25,10 +25,12 @@ let row_focus = 0;
+// not done because there will always be confusion that if action
+// buttons should be shown for (keyboard) focused table row or (mouse) hovered
+// table row.
+const MAX_SELECTABLE_COLS = 2;
```
After playing with this, I feel like having it share space kinda messes up the keyboard navigation.  And I don't like being unable to see the timestamp for the thing I'm looking at.  What do you think about just having these elements by in the `topic` column right-aligned?  That would save space (since most of the space these icons take is their heading at present).

timabbott replied:
(I also don't like losing access to the exact time as you do in a later commit).

timabbott commented:
Are the vim hotkeys changes safe to merge?  If so can you move them to the start of the commit?  Those generally lgtm and we can use them to shrink this branch.

amanagr commented on static/styles/zulip.scss:
```diff
@@ -284,7 +284,9 @@ p.n-margin {
     width: 100%;
     height: 100%;
-    overflow-y: scroll;
+    // We don't want to show scroll bar unless,
```
@timabbott don't merge this yet, It throws off the carefully crafted alignments in various places.

amanagr replied:
Removed this commit, I will come back to this at some other time maybe.

amanagr commented on frontend_tests/node_tests/list_render.js:
```diff
@@ -10,6 +10,7 @@ function Element() {
     return { };
 }
 set_global('Element', Element);
+set_global('ui', {});
```
@sahil839 I cherry-picked our combined effort at fixing this.

sahil839 replied:
Okay, then I will make separate PR for fixing other lists.

amanagr commented:
@timabbott this is ready for review. I can splitup the last commit, if you find it hard to review. I am exhausted for the day. We can test deploy this atleast on czo, if the changes look good to you UI/UX wise.

timabbott commented on static/js/list_render.js:
```diff
@@ -264,7 +264,11 @@ exports.create = function ($container, list, opts) {
+            meta.scroll_container = ui.get_scroll_element(opts.simplebar_container);
+        } else {
+            meta.scroll_container = scroll_util.get_list_scrolling_container($container);
+        }
```
@andersk what are your thoughts on this commit?  I can see how explicit might be better than implicit, but I'm uncertain that this is the right strategy for being explicit.

andersk replied:
`list_render` already has too many variants, do we have to add more? I would think that we either want SimpleBar on all of our `list_render` instances or none of them.

timabbott replied:
Yeah, I sorta feel the same way.  I'm not aware of `list_render` instances where we don't want simplebar, though I suppose it could be annoying to audit them all.

In any case @amanagr can you pull this commit out to a separate PR?

amanagr replied:
I pushed to commit in https://github.com/zulip/zulip/pull/15673 @sahil839 FYI

amanagr replied:
I left the commit here in case you want to test-deploy.

amanagr commented:
Resolved conflict.

amanagr commented:
@timabbott updated!

timabbott commented:
I guess this needs the rebase past `prettier`; make sure you use the process mentioned in Zulip that makes it easy :)

amanagr commented:
updated!

timabbott commented:
Test-deployed on chat.zulip.org.

timabbott commented:
I merged the first half of this PR; I'm going to undeploy this from chat.zulip.org just to make sure we're testing what's in master for a bit.

timabbott commented on static/js/hashchange.js:
```diff
@@ -114,15 +117,6 @@ function do_hashchange_normal(from_reload) {
-            break;
-        case "#search-operators":
-            info_overlay.show("search-operators");
-            break;
```
I kept the `case` statements only here when merging since they provide some protection from future bugs.  (Just not the `info_overlay` lines -- i.e. these now do the same assert as `#drafts` (below))

amanagr replied:
Okay!

timabbott commented on static/js/recent_topics.js:
```diff
@@ -29,6 +29,9 @@ let col_focus = 1;
 const MAX_SELECTABLE_COLS = 4;
+// We don't set a current_focus_element unless we are sure that
+// user is using keyboard to navigate.
+let user_uses_keyboard = false;
```
Can you explain this change in more detail?  I don't understand its purpose.

amanagr replied:
Yeah, If user is not using keyboard to navigate we don't need to call `revive_current_focus` every time we update the view and highlight the focused element. I thought it would stop this unnecessary call for us in case user is just using mouse, I can remove it if it seems odd.

timabbott commented on static/js/recent_topics.js:
```diff
@@ -72,13 +72,17 @@ function revive_current_focus() {
+        current_focus_elem.focus();
+    } else {
+        $("#recent_topics_search").focus();
+>>>>>>> recent_topics: Set default focus to Participated button.
```
Merge conflicts needs fixing :)

amanagr replied:
Oops! Sorry!

timabbott commented on static/styles/night_mode.scss:
```diff
@@ -436,6 +436,10 @@ on a dark background, and don't change the dark labels dark either. */
 
+    #recent_topics_table .unread_topic {
+        background-color: hsl(211, 28%, 16%);
+    }
```
I'm tempted to drop this commit since I found it sorta confusing.

amanagr replied:
Sure.

timabbott commented on static/js/narrow.js:
```diff
@@ -776,8 +776,11 @@ exports.deactivate = function () {
+    // Both All messages and Recent Topics have `undefiend` filter.
+    // Return if already in the All message narrow.
+    if (narrow_state.filter() === undefined && !coming_from_recent_topics) {
         return;
```
I think the `coming_from_recent_topics` check is messy; we should instead check if `current_msg_list === message_list.home`, I think?  I guess I'm not sure what `current_msg_list` should be when we're looking at "Recent topics"; should it be `undefined`?

amanagr replied:
current_msg_list will be `undefined` if user opens czo/#recent_topics first.

amanagr commented:
@timabbott updated according to the discussion in czo.

timabbott commented:
@amanagr can you rebase this?  It's a priority for our 4.0 release so we should try to get this into a mergable state.

amanagr commented:
@timabbott this is ready for review.

timabbott commented:
I merged a couple prep commits.  For this one:

>    recent_topics: Only set deafult focus if user is using keyboard.

Can you fix the typo and also add more to the commit message and/or comments explaining what problem this is solving?

timabbott commented:
This had a weird webpack error when I tried to test-deploy it.
```
+ ./tools/webpack --quiet
Child frontend:
    
    ERROR in ./static/styles/recent_topics.css
    Module build failed (from /srv/zulip-npm-cache/743fdca7dc229079486f65fec679d296488ad38a/node_modules/mini-css-extract-plugin/dist/loader.js):
    ModuleBuildError: Module build failed (from /srv/zulip-npm-cache/743fdca7dc229079486f65fec679d296488ad38a/node_modules/postcss-loader/dist/cjs.js):
    SyntaxError
    
    (65:16) Unknown word
    
      63 |         .table_fix_head {
      64 |             padding: 0 !important;
    > 65 |             // 100px = space occupied by `recent_topics_filter_buttons`( ~49px)
         |                ^
      66 |             //         + give user some extra space at the bottom so that last
      67 |             //           topic row is clearly visible.
    
        at /srv/zulip-npm-cache/743fdca7dc229079486f65fec679d296488ad38a/node_modules/webpack/lib/NormalModule.js:316:20
        at /srv/zulip-npm-cache/743fdca7dc229079486f65fec679d296488ad38a/node_modules/loader-runner/lib/LoaderRunner.js:367:11
        at /srv/zulip-npm-cache/743fdca7dc229079486f65fec679d296488ad38a/node_modules/loader-runner/lib/LoaderRunner.js:233:18
        at context.callback (/srv/zulip-npm-cache/743fdca7dc229079486f65fec679d296488ad38a/node_modules/loader-runner/lib/LoaderRunner.js:111:13)
        at Object.loader (/srv/zulip-npm-cache/743fdca7dc229079486f65fec679d296488ad38a/node_modules/postcss-loader/dist/index.js:101:7)
```

andersk commented:
The hover text for “All messages” still says “All messages (Esc)”, and the the keyboard shortcuts help box still says “Narrow to all unmuted messages: `Esc` or `Ctrl`+`[`”.

If we’re stealing Esc, we should come up with a new shortcut for “All messages”.

andersk commented:
If I’m in a view with messages, press `Esc` to go to recent topics, and then press `Alt`-`←` to go back, the `↑`, `↓`, 'K', 'J' shortcuts to move between messages stop working.

amanagr commented:
> If I’m in a view with messages, press `Esc` to go to recent topics, and then press `Alt`-`←` to go back, the `↑`, `↓`, `K`, `J` shortcuts to move between messages stop working.

I am not able to reproduce this. Can you consistently reproduce it?

andersk commented:
Hmm, I had trouble reproducing from a fresh load, but I found a modified recipe that seems to reproduce consistently:

* Go to http://localhost:9991/.
* Click on the “Search stream / topic” box.
* Press `Alt`-`←` to go back.
* Try `↑`, `↓`, `K`, `J`.

amanagr commented:
Thanks! I can now reproduce this on Firefox. Will try to debug.

amanagr commented:
Fixed. I will work on making recent topics the default view next.

amanagr commented:
@timabbott this is ready for another test deploy.

timabbott commented:
Deploying now.

amanagr commented:
Added another commit to fix a bug reported by @ryanreh99 .

ryanreh99 commented on static/js/hotkey.js:
```diff
@@ -471,7 +471,8 @@ exports.process_hotkey = function (e, hotkey) {
                 !popovers.any_active() &&
-                !overlays.is_active()
+                !overlays.is_active() &&
+                !$("#searchbox_form #search_query").is(":focus")
```
Since this bug also occurs for `stream-list-filter` and `user-list-filter` searchboxes.
Maybe we could update this condition to check if `document.activeElement.tagName` is not "input".

amanagr replied:
Weird! I was not able to reproduce this for stream and user search. Are you able to reproduce it consistently? I am actually confused here a bit since when stream / user search is active `popovers.any_active()` is true here, I will try debugging more on this.

amanagr replied:
Apparently `$("[class^='column-'].expanded").length` is 1 when user / stream search is active, hence popovers.any_active is true. So, ideally this bug should not be present in user / stream search. Can you please try reproducing on this branch and report?

ryanreh99 replied:
Hmm, I can't reproduce this consistently.
Occurs after a few tries, but once it occurs I can reproduce always.
This was on CZO. Will test on this branch and try to find a reproducer.

amanagr replied:
Since this could be a flaky reproduce, I will add a fix for stream / user search explicitly anyway. Thanks!

amanagr commented:
@timabbott this is probably ready for another test deploy, I have added a lot of changes since last deploy. I still need to fix the issue I noted in https://chat.zulip.org/#narrow/stream/2-general/topic/Recent.20Topics/near/1023348 though.

timabbott commented:
Redeploying now.

amanagr commented:
@timabbott I have fixed all the bugs that were reported or I noticed.

timabbott commented:
Deploying on chat.zulip.org now.

timabbott commented:
Are there chunks of this that could be merged independently without doing the cutover to the new "Recent topics" model?

amanagr commented:
Fix 6 commits can be merged. The rest are mostly fixes to bugs which happened after we moved RT away from overlay. So, it makes sense to merge them afterwards (even though we can merge them beforehand).

amanagr commented on templates/zerver/app/left_sidebar.html:
```diff
@@ -5,7 +5,7 @@
                 <a href="#all_messages" class="home-link">
                     <span class="filter-icon">
-                        <i class="fa fa-home" aria-hidden="true"></i>
+                        <svg class="globe-web-icon" viewBox="0 0 495.9 496"><path d="M336.5,160C322,70.7,287.8,8,248,8s-74,62.7-88.5,152ZM152,256a626.79,626.79,0,0,0,3.3,64H340.6a626.79,626.79,0,0,0,3.3-64,626.79,626.79,0,0,0-3.3-64H155.3A626.79,626.79,0,0,0,152,256Zm324.7-96a248.52,248.52,0,0,0-158-141.6c24.4,33.8,41.2,84.7,50,141.6ZM177.2,18.4A248.36,248.36,0,0,0,19.3,160h108c8.7-56.9,25.5-107. ...
```
![image](https://user-images.githubusercontent.com/25124304/95212704-9d339d00-080b-11eb-9066-1732afe22391.png)
<img width="790" alt="Screenshot 2020-10-06 at 7 32 26 PM" src="https://user-images.githubusercontent.com/25124304/95212737-a3297e00-080b-11eb-95bf-bf6bb9c8c7b4.png">

timabbott replied:
Looks nice visually; do we need to also change something in the "Search" UI?

timabbott commented:
I merged a few of the early commits as the series ending with 623fb634061ccd52032acba522f73ff7a5070963.  A few others need a rebase.

amanagr commented:
Last four commits.
Commit 1:
<img width="815" alt="Screenshot 2020-11-04 at 9 51 23 AM" src="https://user-images.githubusercontent.com/25124304/98069517-fcc1ae80-1e84-11eb-9ff1-556f2dbfd65b.png">
Commit 2:
<img width="815" alt="Screenshot 2020-11-04 at 9 51 31 AM" src="https://user-images.githubusercontent.com/25124304/98069549-106d1500-1e85-11eb-93fc-b30da856c130.png">
Commit 3:
<img width="815" alt="Screenshot 2020-11-04 at 9 51 38 AM" src="https://user-images.githubusercontent.com/25124304/98069579-1ebb3100-1e85-11eb-88e6-4afd5bb5c94e.png">
Commit 4:
<img width="815" alt="Screenshot 2020-11-04 at 10 01 31 AM" src="https://user-images.githubusercontent.com/25124304/98069608-2d094d00-1e85-11eb-93ed-34a18868efac.png">

YashRE42 commented:
small nitpick here but if we change the icon in the left side bar, we should change it in the narrow_view_header as well (top navbar).

amanagr commented:
<img width="610" alt="Screenshot 2020-11-10 at 2 25 17 PM" src="https://user-images.githubusercontent.com/25124304/98651765-ab239300-2360-11eb-9f91-361344ef1cda.png">

Changed all messages icon.

amanagr commented:
The build is failing due to some urls missing from the stable version of zulip readthedocs.

amanagr commented:
@timabbott this is ready for redeploy.

timabbott commented:
@amanagr can your rebase this?

timabbott commented on static/js/narrow.js:
```diff
@@ -784,7 +788,9 @@ exports.deactivate = function () {
      */
     search.clear_search_form();
-    if (narrow_state.filter() === undefined) {
+    // Both All messages and Recent Topics have `undefiend` filter.
```
"undefined" is mispelled.

timabbott commented on static/js/hashchange.js:
```diff
@@ -133,14 +148,18 @@ function do_hashchange_normal(from_reload) {
         case "#streams":
         case "#organization":
         case "#settings":
-        case "#recent_topics":
```
There's a broken intermediate state in here, where various other items have the `recent_topics` code running.

timabbott replied:
Fixed in the commit where it's introduced and resolved the merge conflict when you fix it ~4 commits later :)

timabbott commented on static/js/hotkey.js:
```diff
@@ -272,7 +273,7 @@ exports.process_escape_key = function (e) {
     }
 
-    hashchange.go_to_location("");
+    hashchange.go_to_location("#");
```
Why does this line change?

amanagr replied:
I don't remember. I can see how this change doesn't belong here. Removed.

timabbott commented on static/js/hotkey.js:
```diff
@@ -476,7 +477,16 @@ exports.process_hotkey = function (e, hotkey) {
+        case "open_recent_topics":
+        case "escape":
+            if (
+                ["#recent_topics", "#", ""].includes(window.location.hash) &&
```
We should abstract this first check to a function with a name exported by `hashchange.js`.

amanagr replied:
Okay, Opened #17330 .

timabbott commented:
OK, I think this is pretty close to mergeable.  Here's the things I changed in addition to the comments I noted and resolved above:

* [x] For `recent_topics: Fix % of space taken by columns.`, the commit message should be copied into the comment
* [x] The `templates/zerver/portico-header-dropdown.html` changes replacing `fa-home` seem wrong to me, since that context is the mobile menu for portico pages (I think).

Here's a list of things we should talk about and potentially fix before the next release, but don't seem like a blocker for this PR:
* [ ] The CSS for this makes a lot of use of `!important`; I'd like to make sure that's actually necessary as generally we try to avoid using that directive in the codebase when not required.
* [ ] I think we should create a more declarative exported function for "Messages not visible in center pane" that handles overlays, "recent topics", and anything of that form we might use in the future, that can be called from places like `hotkeys.js`.  We should keep in mind that anything we do in `hotkey.js` needs to be fast if it's a code path before we've decided we're not in the compose box (and other input elements), to avoid slowing down typing actual messages.
* [ ] We probably will also want to think about simplifying the `hotkey.js` list of exceptions for `recent_topics` -- it's generally undesirable to have a list of specific widgets where we shouldn't eat input; better would be to only eat input relevant to the target widget.  E.g. I think if we added a new text box in a sidebar, we'd need to remember to adjust that list.
* [ ] How did you pick the key `t` for "Go to search box" within "recent topics"?
* [ ] We will probably want to rename the `#all_messages` URL to something more similar to other message views.  Probably a broader discussion because we may want to bundle that into a larger/deferred plan to e.g. replace `/#narrow/` with `/messages/` and/or other migrations to avoid churning what users see too much; I'm not sure.  Start a chat.zulip.org thread about naming as the next step for this one.
* [ ] From recently playing with "recent topics", I really want to be able sort by number of unreads.
* [ ] Why does tab_pane lose the `active` class?
* [ ] For the `thead .last_msg_time_header` logic, how was the 93px number computed?
* [ ] Can you explain the bug being fixed `recent_topics: Fix partially hidden row sometimes in recent topics` more clearly in the commit message and/or comment?  I feel like that line is going to be something we're asking ourselves if we can remove down the line.
* [ ] For issues like `MessageListView: Don't reselect message_id if RT is open.`, should we instead by making `current_message_list` undefined when showing "Recent topics"?  That's likely a bigger refactor since we'd need to audit for what changing that invariant might impact, but that might be more correct logically.  Until then, I think the current code will DTRT in all existing cases, so it's mergeable, just a potential liability.
* [ ] `recent_topics_row: Display stream icon inline with wrapped text`: Can we change the HTML to a single div with two classes, rather than two divs around each other?
* [ ] The `resize_stream_filters_container` call seems hacky; I kinda feel like that should live inside `pm_list.close()` instead; needs investigation.
* [ ] https://github.com/zulip/zulip/pull/15650#pullrequestreview-591728883
* [ ] https://github.com/zulip/zulip/pull/15650#discussion_r577226127

I'm going to hold off on merging this right now since a few of these comments are about commit messages, but please post diffs for any commits you edit after this point, as I hope to not read the original commit series again before merging.

andersk commented:
At some point we need to change the HTML `<title>` of the All messages view (currently “home”) and of the Recent topics view (currently “Recent topics (beta)”).

amanagr commented:
I will followup on the comments at my own pace and let you know when this is ready for review again.

amanagr commented:
For things I will followup on / create issues after this PR is merged, I am tracking them here - https://github.com/amanagr/zulip/projects/2#column-12991478

> The CSS for this makes a lot of use of !important; I'd like to make sure that's actually necessary as generally we try to avoid using that directive in the codebase when not required.

I looked into them. I think I added them initially only because they were required.

> I think we should create a more declarative exported function for "Messages not visible in center pane" that handles overlays, "recent topics", and anything of that form we might use in the future, that can be called from places like hotkeys.js. We should keep in mind that anything we do in hotkey.js needs to be fast if it's a code path before we've decided we're not in the compose box (and other input elements), to avoid slowing down typing actual messages.

> We probably will also want to think about simplifying the hotkey.js list of exceptions for recent_topics -- it's generally undesirable to have a list of specific widgets where we shouldn't eat input; better would be to only eat input relevant to the target widget. E.g. I think if we added a new text box in a sidebar, we'd need to remember to adjust that list.

Agreed that we need to think about both the above things. Probably the first one will simplify the other. I will followup on these personally soon.


> How did you pick the key t for "Go to search box" within "recent topics"?

To simplify the finger movement to go to RT search. You just tap `t` twice and you are in RT search from any narrow.



>  From recently playing with "recent topics", I really want to be able sort by number of unreads (at least with my current data set with 1500 unread messages -- basically it'll let me quickly find the longest threads I'm not caught up on).

#17347


> Why does tab_pane lose the active class?

The `active` class make the tab-pane visible on initial page load, which is no longer the case since RT is the default view now.

> For the thead .last_msg_time_header logic, how was the 93px number computed?

It is the width occupied by `Last message + icon` computed by chrome.

> Can you explain the bug being fixed recent_topics: Fix partially hidden row sometimes in recent topics more clearly in the commit message and/or comment? I feel like that line is going to be something we're asking ourselves if we can remove down the line.

Okay, edited the commit message to be more verbose.

> For issues like MessageListView: Don't reselect message_id if RT is open., should we instead by making current_message_list undefined when showing "Recent topics"? That's likely a bigger refactor since we'd need to audit for what changing that invariant might impact, but that might be more correct logically. Until then, I think the current code will DTRT in all existing cases, so it's mergeable, just a potential liability.

Agreed, I will followup on it.

> recent_topics_row: Display stream icon inline with wrapped text: Can we change the HTML to a single div with two classes, rather than two divs around each other?

Okay done! 4699b0e5953209e01ebade095feb34808b75a28e

> The resize_stream_filters_container call seems hacky; I kinda feel like that should live inside pm_list.close() instead; needs investigation.

Actually, pm_list.close() is called only with handle_narrow_activated/deactivated. And `resize_stream_filters_container` is called after this function is called. I agree that it seems hacky, I will take a hard look at it again.

> We should audit our documentation to make sure we discuss Recent Topics enough, and adjust anything talking about "All messages" as the default view.

Done e073525a2379f9d0122a770618a62688253d69b5 & the second one looks like a big task in itself will followup on it soon. - https://github.com/amanagr/zulip/projects/2#card-55252866

> We should do testing for how this view works for users creating a new account in a large organization like chat.zulip.org. I think ONBOARDING_TOTAL_MESSAGES being 1000 means that users should see the last 1K messages in their default streams in "Recent topics", and that's probably enough history to be able to scroll to the bottom. But we should test, and maybe think about having text when you hit the bottom of the table that makes it possible to discover that "Recent topics" only provides limited history from before you joined. (We may later want to have some way to "Fetch more from before one joined" using streams:public type narrows, but that's certainly a flourish we don't need to block on). Because we'll be dropping new users into this view pretty quickly, we want to make sure it doesn't make them think it's impossible to read messages older than 2 days ago in a community like chat.zulip.org that might do 1K messages in 2 days.

Okay will followup on this.

> I kinda want "Recent topics" to be the top item on the left sidebar, but I think it might end up messing up the conceptual structure (where having the "All messages" unread count at the top signals clearly that it's the grand total).

Started discussion about it in #frontend.

amanagr commented:
> At some point we need to change the HTML <title> of the All messages view (currently “home”) and of the Recent topics view (currently “Recent topics (beta)”).

Done 746d7db63e3a1d735c693640689f562c8ef09b3b

amanagr commented:
I am working on fixing puppeteer tests. Will fix and push them soon.

amanagr commented:
@timabbott see bed52e4880ba7950bdad4b94544192c7997315e3 as per our discussion.

amanagr commented:
Production tests are failing due to network failure.

timabbott commented on static/js/recent_topics.js:
```diff
@@ -484,7 +484,7 @@ exports.show = function () {
     compose_actions.cancel();
 
-    narrow.narrow_title = "Recent topics (beta)";
+    narrow.narrow_title = "Home";
```
I think we might want this still to be called "Recent topics".

timabbott commented on templates/zerver/help/reading-strategies.md:
```diff
@@ -25,6 +25,22 @@ only going to learn one way to do it, this is the one.
+
+* There filters at the top can help you quickly find the conversations
+  that are specifically relevant to you.  Filtering for conversations
+  with unread messages that you've participated is often useful.
```
"that you've participated _in_ is often useful"

timabbott commented:
Merged after fixing those last two items!  Huge thanks for all your work on this.

I'm deploying master to chat.zulpi.org shortly, and have opened https://github.com/zulip/zulip/issues/17415 to track future work.
